src/core/kconfigelektra.cpp

```diff
@@ -0,0 +1,225 @@
+//
+// Created by felix on 28.10.19.
+//
+
+#include "kconfigelektra.h"
+#include "kconfigdata.h"
+
+#include <utility>
+#include <iostream>
+
+using namespace kdb;
+
+KConfigElektra::KConfigElektra(std::string appName, uint majorVersion, std::string profile) : app_name(std::move(
+        appName)), major_version(majorVersion), profile(std::move(profile)) {}
+
+KConfigElektra::KConfigElektra(std::string appName, uint majorVersion) :
+        KConfigElektra::KConfigElektra(std::move(appName), majorVersion, "current") {}
+
+KConfigElektra::~KConfigElektra() = default;
+
+/**
+ * Move both name iterators to the point, where they differ.
+ *
+ * @param parent the iterator of the parent key
+ * @param child the iterator of the child key
+ */
+inline void traverseIterators(Key::iterator* parent, Key::iterator* child) {
+    /*
+     * First item of iterator seems to be namespace, as namespaces can differ with cascading keys, we skip these.
+     * (Loop initializer)
+     *
+     * Loop as long as the paths are the same and iterate through the path segments.
+     */
+    for ((*parent)++, (*child)++; parent->get() == child->get(); (*parent)++, (*child)++);
+}
+
+struct KConfigKey {
+    std::string group;
+    std::string key;
+};
+
+inline KConfigKey elektraKeyToKConfigKey(Key::iterator* key, Key::iterator* end) {
+    std::ostringstream group;
+    std::string key_name;
+    bool first = true;
+
+    for(; (*key) != (*end); (*key)++) {
+        if (!key_name.empty()) {
+            if (!first) {
+                group << "\x1d";
+            } else {
+                first = false;
+            }
+            group << key_name;
+        }
+        key_name = key->get();
+    }
+
+    return KConfigKey {
+        group.str(),
+        key_name,
+    };
+}
+
+KConfigBackend::ParseInfo
+KConfigElektra::parseConfig(const QByteArray& /*locale*/, KEntryMap &entryMap, KConfigBackend::ParseOptions options) {
+    //TODO error handling
+    //TODO properly handle parse options (or figure out how they actually work)
+
+    KDB kdb = KConfigElektra::open_kdb();
+    KeySet keySet;
+
+    bool oGlobal = options & ParseGlobal;
+
+    Key parent_key = Key(app_key());
+
+    kdb.get(keySet, parent_key);
+    keySet = keySet.cut(parent_key);
+
+    Key key;
+
+
+    while((key = keySet.pop()) != nullptr) {
```

Comment:
can `for(auto& key : keySet)` be used here?

Reply:
I had to change it to `const` but it works (`for(const auto& key : keySet)`)